nodego/supervisor.go

```diff
@@ -0,0 +1,189 @@
+package nodego
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"runtime/debug"
+	"strconv"
+	"sync"
+	"time"
+)
+
+type logEntry struct {
+	TextPayload string
+	Severity    string
+	Time        string
+	ExecutionID string
+}
+
+func (e *logEntry) ConsoleOutput() []byte {
+	var logBuf bytes.Buffer
+	fmt.Fprintf(&logBuf, "[%s]", e.Severity[:1])
+	fmt.Fprintf(&logBuf, "[%s]", time.Now().Format(isoTimeFormat))
+	if functionExecutionId != "" {
+		fmt.Fprintf(&logBuf, "[%s]", functionExecutionId)
+	}
+	logBuf.WriteRune(' ')
+	logBuf.WriteString(e.TextPayload)
+	if e.TextPayload[len(e.TextPayload)-1] != '\n' {
+		logBuf.WriteRune('\n')
+	}
+	return logBuf.Bytes()
+}
+
+type logBatch struct {
+	Entries []*logEntry
+
+	payloadLength int
+	ready         chan struct{}
+}
+
+func (batch *logBatch) AddEntry(entry *logEntry) {
+	if batch.Entries == nil {
+		close(batch.ready)
+	}
+	batch.Entries = append(batch.Entries, entry)
+	batch.payloadLength += len(entry.TextPayload)
+}
+
+var (
+	currentLogBatch      *logBatch
+	currentLogBatchMutex sync.Mutex
+	logBatchesToReport   = make(chan *logBatch, 5)
+)
+
+var (
+	functionExecutionId string
+)
+
+var (
+	infoLogWriter  supervisorWriter = "INFO"
+	errorLogWriter supervisorWriter = "ERROR"
+
+	// InfoLogger is a logger that batches sends logs to the supervisor with a
+	// severity level of INFO
+	InfoLogger = log.New(infoLogWriter, "", 0)
+	// InfoLogger is a logger that batches sends logs to the supervisor with a
+	// severity level of ERROR
+	ErrorLogger = log.New(errorLogWriter, "", 0)
+)
+
+func init() {
+	if supervisorHostname != "" && supervisorInternalPort != "" {
+		startNewLogBatch()
+		go logBatchReportWorker()
+	}
+}
+
+func startNewLogBatch() *logBatch {
+	currentLogBatch = &logBatch{
+		ready: make(chan struct{}),
+	}
+	logBatchesToReport <- currentLogBatch
+	return currentLogBatch
+}
+
+type supervisorWriter string
+
+const isoTimeFormat = "2006-01-02T15:04:05.999Z07:00"
+
+func (w supervisorWriter) Write(p []byte) (n int, err error) {
+	entry := &logEntry{
+		TextPayload: string(p),
+		Severity:    string(w),
+		Time:        time.Now().Format(isoTimeFormat),
+	}
+	if functionExecutionId != "" {
+		entry.ExecutionID = functionExecutionId
+	}
+	currentLogBatchMutex.Lock()
+	defer currentLogBatchMutex.Unlock()
+	if currentLogBatch == nil {
+		return os.Stderr.Write(entry.ConsoleOutput())
+	}
+	// Start a new batch if the current one would grow too much.
+	if len(currentLogBatch.Entries)+1 > maxLogBatchEntries || currentLogBatch.payloadLength+len(entry.TextPayload) > maxLogBatchLength {
+		startNewLogBatch()
+	}
+	currentLogBatch.AddEntry(entry)
+	return len(entry.TextPayload), nil
+}
+
+func logBatchReportWorker() {
```

Comment:
Optional: It might be nice to have a ticker or something that adds a time trigger on log buffering in addition to the number/length based trigger you currently have. Maybe add a TODO?